pkg/lib/bad/error.go

```diff
@@ -0,0 +1,184 @@
+type Error struct {
+	// The type of the error. For native errors, this is just the error text,
+	// but more structured errors this is a unique identifier that clients can
+	// use to display certain types of important errors in specific ways.
+	// Essentially, the "type" identifies the type of thing in the "data" field,
+	// and allows the client to unmarshall the data field appropriately.
+	//
+	// For example, if a client knows how to represent errors about resource
+	// limits being unmet, the type may be "Unmet resource limits" and the data
+	// field may include structured information about what resource limits are
+	// being unmet that the client can choose to display in a number of ways.
+	Type string `json:"type"`
+	Data any    `json:"data"`
+
+	// The subject of the error, which identifies where the problem occurred and
+	// is used to select appropriate status codes and retry information.
+	//
+	// Not every Error needs to have a Subject specified – instead subjects can
+	// be added to tag any error from a certain place. For example, a Validate()
+	// method will almost always be returning an error due to user input, so any
+	// error emitted from this place can be wrapped with a ErrSubjectInput.
+	Subject ErrorSubject `json:"subject"`
+
+	// Any child errors that this error wraps.
+	Errs []Error `json:"errs"`
+}
+
+// Error() is necessary for Error to implement the error interface, which makes
+// it much easier to pass back from functions. We shouldn't actually use this
+// except in internal logging, but it implements a simple printer that will
+// output the entire tree of errors, skipping any blank levels.
+func (err *Error) Error() string {
+	var out strings.Builder
+	if err == nil {
+		return ""
+	}
+
+	indent := ""
+	if len(err.Type) > 0 {
+		out.WriteString("* ")
+		out.WriteString(err.Type)
+		indent = "\t"
+		if len(err.Errs) > 0 {
+			out.WriteString(": ")
+		}
+		out.WriteRune('\n')
+	}
+	for _, subError := range err.Errs {
+		for _, line := range strings.SplitAfter(subError.Error(), "\n") {
+			if line == "" {
+				continue
+			}
+			out.WriteString(indent)
+			out.WriteString(line)
+		}
+	}
+	return out.String()
+}
+
+var _ error = (*Error)(nil)
+
+// Leaves returns just the Errors from the tree that do not have any children.
+// These Errors are normally the ones with the most specific error information.
+// The order in which the leaves are returned is unspecified.
+func (err *Error) Leaves() []Error {
+	if err == nil {
+		return nil
+	}
+
+	if len(err.Errs) == 0 {
+		return []Error{*err}
+	}
+
+	leaves := make([]Error, 0, len(err.Errs))
+	for _, subErr := range err.Errs {
+		leaves = append(leaves, subErr.Leaves()...)
+	}
+	return leaves
+}
+
+// Input transforms the error to report that it was the result of user input not
+// meeting requirements. Data is an optional parameter that should be specified
+// 0 or 1 times and will replace any data in the error if specified.
+func Input(err error, data ...any) error {
```

Comment:
(I think) This is variadic to allow for optional values - we can pass nothing or something. In the event that more than one value is provided, which readers of the method declaration will believe to be acceptable at first glance, the method will panic. We can't have this type of behavior in our code.

Reply:
This is the pattern used by other packages such as echo for handling optional parameters: https://github.com/labstack/echo/blob/447c92d842e2f91c07eb032f619bea212beeee60/echo.go#L892-L899 Yes, it does require the user to actually read the comment or to test their code. 

In any case, it sounds like we maybe don't want this function and/or an option fn may be more useful.

Reply:
The only behavior I take issue with is panicking.

 I would rather this method drop any extra fields, as echo does, or we could add a new metho - `InputWiffMsg` - that accepts a single field. functional options are, err, an option as well - but I worry they will be a bit verbose to write.